### PR TITLE
Project task2

### PR DESCRIPTION
**Pull request by CuteKirichenko**
Validators

### CODE REVIEW

File: project/src/main/java/com/kirichenko/App.java [hunk 1]
```diff
@@ -17,6 +17,8 @@
 public class App {
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/App.java]: First of all, you should always check unused imports and classes. If you are not going to use them then delete them.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 2]
```diff
@@ -1,13 +1,53 @@
...
 
+import com.kirichenko.dao.ClientDao;
 import com.kirichenko.exceptions.BusinessException;
 import com.kirichenko.validators.ValidationService;
 
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 public class ValidationServiceImpl implements ValidationService {
+    private ClientDao clientDao;
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: unused field

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 3]
```diff
@@ -1,13 +1,53 @@
...
+    private final static String patternMail = "^[A-Za-z0-9](.*)([@]{1})(.+)(\\.)(.+)";
+    private final static String patternPhone = "\\+38(-\\d{3}){2}(-\\d{2}){2}";
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: this regex will allow a lot of figure combinations. It checks that there are figures and dashes between groups of them.
+38-000-000-0000 will be a valid number
You can use regexes for such tasks but usually, it's quite tricky and error-prone.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 4]
```diff
@@ -1,13 +1,53 @@
...
+    public void validatePhone(String phone) throws BusinessException {
+        Matcher matcher = phonePattern.matcher(phone);
+        boolean b = matcher.matches();
+        if (!b) {
+            throw new BusinessException("Incorrect phone!");
+        }
+    }
+
+    @Override
+    public void validatePhoneDuplicate(String phone) throws BusinessException {
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: this check could be done in the validatePhone method

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 5]
```diff
@@ -1,13 +1,53 @@
...
+        long id = 0;
+        while (clientDao.findClient(id) != null) {
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: It's totally normal to have gaps in the id sequence. So, it's highly likely that this cycle doesn't check all the clients. Actually, you have a method in your service that is supposed to return all the clients.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 6]
```diff
@@ -1,13 +1,53 @@
...
 
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 public class ValidationServiceImpl implements ValidationService {
+    private ClientDao clientDao;
+    private final static String patternMail = "^[A-Za-z0-9](.*)([@]{1})(.+)(\\.)(.+)";
+    private final static String patternPhone = "\\+38(-\\d{3}){2}(-\\d{2}){2}";
+    private Pattern mailPattern = Pattern.compile(patternMail);
+    private Pattern phonePattern = Pattern.compile(patternPhone);
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: these fields can be constants and you can inline your regex strings and remove String fields at all

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java [hunk 7]
```diff
@@ -1,13 +1,53 @@
...
+        if (!b) {
+            throw new BusinessException("Incorrect phone!");
+        }
+    }
+
+    @Override
+    public void validatePhoneDuplicate(String phone) throws BusinessException {
+        long id = 0;
+        while (clientDao.findClient(id) != null) {
+            if (phone.equals(clientDao.findClient(id).getPhone())) {
```
Comment (YuriHeiko) [inline project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: DB calls are expensive operations. You make the same call twice. Instead, you can use a variable here and there and reduce db calls

File: project/src/main/java/com/kirichenko/App.java (hunk 1 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/App.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 3 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Simplify regex for phone.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 4 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 2 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Without it, validatePhoneDuplicate can''t work.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 6 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 7 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 5 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: I have changed logic.

File: project/src/main/java/com/kirichenko/App.java (hunk 1 above)
Comment (YuriHeiko) [reply project/src/main/java/com/kirichenko/App.java]: Check OrderService

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 2 above)
Comment (YuriHeiko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Please check where and how you assign it

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 3 above)
Comment (YuriHeiko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Saying that "+38-000-000-0000 will be a valid number" I meant that it is wrong and you need not simplify it but fix

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 4 above)
Comment (YuriHeiko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: This method can be private now

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 6 above)
Comment (YuriHeiko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Please figure out how constants(public static final fields) should be named in Java

File: project/src/main/java/com/kirichenko/App.java (hunk 1 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/App.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 4 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Fixed.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 6 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Refactored.

File: project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java (hunk 2 above)
Comment (CuteKirichenko) [reply project/src/main/java/com/kirichenko/validators/impl/ValidationServiceImpl.java]: Use the same way as ClientServiceImpl (constructor).

**Pull request merged**